core/indexer/interface.go

```diff
@@ -2,44 +2,70 @@ package indexer
 
 import (
 	"bytes"
-	"io"
-
 	"github.com/ElrondNetwork/elrond-go/core/statistics"
 	"github.com/ElrondNetwork/elrond-go/data"
 	"github.com/ElrondNetwork/elrond-go/data/block"
 	"github.com/ElrondNetwork/elrond-go/process"
 	"github.com/elastic/go-elasticsearch/v7/esapi"
+	"io"
+	"time"
 )
 
 // Indexer is an interface for saving node specific data to other storage.
 // This could be an elastic search index, a MySql database or any other external services.
 type Indexer interface {
 	SetTxLogsProcessor(txLogsProc process.TransactionLogProcessorDatabase)
 	SaveBlock(body data.BodyHandler, header data.HeaderHandler, txPool map[string]data.TransactionHandler, signersIndexes []uint64, notarizedHeadersHashes []string)
+	RevertIndexedBlock(header data.HeaderHandler)
 	SaveRoundsInfos(roundsInfos []RoundInfo)
 	UpdateTPS(tpsBenchmark statistics.TPSBenchmark)
 	SaveValidatorsPubKeys(validatorsPubKeys map[uint32][][]byte, epoch uint32)
 	SaveValidatorsRating(indexID string, infoRating []ValidatorRatingInfo)
+	StopIndexing() error
+	GetQueueLength() int
 	IsInterfaceNil() bool
 	IsNilIndexer() bool
 }
 
-// DatabaseHandler is an interface used by elasticsearch component to prepare data to be saved on elasticsearch server
-type DatabaseHandler interface {
+// DispatcherHandler is an interface for the data dispatcher data is used to index data in database
+type DispatcherHandler interface {
+	StartIndexData()
+	Close() error
+	Add(item *workItem)
+	GetQueueLength() int
+	SetTxLogsProcessor(txLogsProc process.TransactionLogProcessorDatabase)
+}
+
+// ElasticIndexer defines the interface for the elastic search indexer
+type ElasticIndexer interface {
```

Comment:
Why do we still need this interface. The package **should** be simplified as much as possible (right now is very hard to create a mental map of all components). Suggestion below:
`Indexer` interface is used by the rest of the project. We should have an implementation, let's call it `elasticIndexer`. This implementation delegates the functionality towards various internal components that implement a couple of interfaces useful in tests through mocks. A factory can be used to assemble a functional elasticIndexer implementation by providing the smaller components used. This can be used for as many levels you want but should be kept as simple as possible, avoiding duplicated code, breaking code in smaller components. Right now it's a complete mess in indexer package.

Reply:
refactored